[C3] fix: use a valid compatibility date for worker templates

petebacondarwin opened the pull request:
Fixes https://github.com/cloudflare/workers-sdk/issues/2385

**What this PR solves / how to test:**

Previously, we changed wrangler.toml to use the current date for the compatibility_date setting in wrangler.toml when generating workers. But this is almost always going to be new recent and results in a warning.

Now we look up the most recent compatibility date via npm on the workerd package and use that instead.

To test, build the create-cloudflare-package and then use it to build a workers project.
Check that the output displays:

```
├ Retrieving current workerd compatibility date 
│ compatibility date 2023-05-18
```

And that the generated wrangler.toml contains this date:

```toml
compatibility_date = "2023-05-18"
```

**Associated docs issue(s)/PR(s):**

- N/A

**Author has included the following, where applicable:**

- [x] Tests
- [x] Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))

**Reviewer is to perform the following, as applicable:**

- Checked for inclusion of relevant tests
- Checked for inclusion of a relevant changeset
- Checked for creation of associated docs updates
- Manually pulled down the changes and spot-tested

petebacondarwin commented on packages/create-cloudflare/src/workers.ts:
```diff
@@ -158,3 +158,28 @@ async function updateFiles(ctx: Context) {
+		startText: "Retrieving current workerd compatibility date",
+		transform: (result) => {
+			try {
+				const date = result.split(".")[1];
+				return `${date.slice(0, 4)}-${date.slice(4, 6)}-${date.slice(6, 8)}`;
```
If `date` is undefined the catch block with handle the throw and result in the default value.
Also the transform function will only be called if the `npm ...` command returns a 0 status code.

So do we need to be more careful with parsing here?

petebacondarwin replied:
Ah!  Good point...

petebacondarwin replied:
Reworked it to handle failure modes a bit better, and even changed the tests to be able to check these.

rozenmd commented on packages/create-cloudflare/src/workers.ts:
```diff
@@ -147,8 +151,8 @@ async function updateFiles(ctx: Context) {
 		.replace(
-			/^compatibility_date = .+$/m,
-			`compatibility_date = "${new Date().toISOString().substring(0, 10)}"`
+			/^compatibility_date\s=\s"<TBD>"$/m,
```
Do you need to update the name's `.replace` too?

petebacondarwin replied:
Good spot. I'll add that.

petebacondarwin replied:
I also noticed that my regex was a bit too restrictive regarding the spaces around the `=`.

rozenmd commented on packages/create-cloudflare/src/helpers/command.ts:
```diff
@@ -223,3 +249,25 @@ export const listAccounts = async () => {
 };
+
+/**
+ * Look up the latest release of workerd and use its date as the compatibility_date
```
Would be good to add a comment about the expected input format or something, like:
"Extracts the latest release date from workerd's version number (of the form 'major.yyyymmdd.patch')"
or something like that

rozenmd reviewed (approved):
Looks good to me, I'd probably wait for a review from someone more familiar with C3 though.

RamIdeas commented on packages/create-cloudflare/src/workers.ts:
```diff
@@ -145,10 +149,10 @@ async function updateFiles(ctx: Context) {
 		.replace(
-			/^compatibility_date = .+$/m,
-			`compatibility_date = "${new Date().toISOString().substring(0, 10)}"`
+			/^compatibility_date\s*=\s*"<TBD>"$/m,
```
If a template add a comment after either the date or name these regexs won't match

Since you're matching the placeholder value and quotes, you can simply replace that and leave the rest of the line intact

petebacondarwin replied:
Agh! Of course. Good call.

RamIdeas commented on packages/create-cloudflare/src/workers.ts:
```diff
@@ -145,10 +149,10 @@ async function updateFiles(ctx: Context) {
 		.replace(
-			/^compatibility_date = .+$/m,
-			`compatibility_date = "${new Date().toISOString().substring(0, 10)}"`
+			/^compatibility_date\s*=\s*"<TBD>"$/m,
```
```suggestion
			/^compatibility_date\s*=\s*"<TBD>"/m,
```

petebacondarwin commented:
OK I have rebased and I think I have addressed all the comments.
My slight worry is that the type parameters are a bit messed up. But it works as is right now. We can probably be more sophisticated later?

Cherry commented on packages/create-cloudflare/src/workers.ts:
```diff
@@ -147,12 +151,14 @@ async function updateFiles(ctx: Context) {
+	if (contents.packagejson.name === "<TBD>") {
+		contents.packagejson.name = ctx.project.name;
+	}
 	contents.wranglertoml = contents.wranglertoml
```
Would a `toml` parser not make more sense here to avoid any quirkiness that comes with regex replacing?

petebacondarwin replied:
A toml parser would be more effective for bizarre templates, but they are not very useful for manlpulating toml, since you lose comments and layout. In these templates, the toml should not be too messy.

Cherry replied:
Comments make a lot of sense! Thanks for the insight.

Cherry commented on packages/create-cloudflare/src/helpers/command.ts:
```diff
@@ -289,3 +314,28 @@ export const listAccounts = async () => {
+ * @returns The latest compatibility date for workerd in the form "YYYY-MM-DD"
+ */
+export async function getWorkerdCompatibilityDate() {
+	return runCommand("npm info workerd dist-tags.latest", {
```
Out of interest, why run a command here instead of just hitting the npm API directly? That seems safer and more reliable to me? 

`https://registry.npmjs.org/workerd`

petebacondarwin replied:
Thanks for the suggestion @Cherry - @mrbbot also suggested this during an in-person review.
 
My take was:
- all these templates will require npm commands to run, so if npm is broken then you have bigger problems, so reliability seems irrelevant
- I am not sure what is meant by safer here?
- It is possible that hitting npm REST directly might be faster, but I doubt that it would be notable in the bigger picture of generating a project

I am also not clear how public and reliable the npm REST API is. I can't find any official (non-archived) docs for the API. Finally, using the CLI directly could benefit from accessing the cached packages (totally guessing here!) which actually might be faster than hitting the API??

So, this is working right now. I am happy to change it if there is enough evidence or concern about this approach.

Cherry replied:
Thanks for the detailed thoughts @petebacondarwin!

My comment about being safer was simply due to avoiding needing to spin up a new process, etc. and the overhead that comes with that. I know it's technically possible to run `node` without `npm` even being installed, but as you noted, almost all of these frameworks and templates recommended/require `npm` in some capacity anyway so this probably isn't a huge concern.

For the npm API, I don't think they have any official public docs outside of the packages they publish to interact with it, such as https://github.com/npm/pacote and https://github.com/npm/npm-registry-fetch, etc. `npm info` also seems to be an alias for `npm view`, and it doesn't seem to use `cache` at all from what I can see at https://github.com/npm/cli/blob/latest/lib/commands/view.js, and just hits the npm API directly using `pacote`.

As you mentioned though, this is probably _fine_ and not a huge concern. 👍

petebacondarwin commented on packages/create-cloudflare/src/helpers/command.ts:
```diff
@@ -79,27 +81,46 @@ export const runCommand = async (
+						if (opts.fallback) {
+							resolve(opts.fallback(e));
+						} else {
+							reject(new Error(output, { cause: code }));
+							reject(e);
```
Oops rebase mess.

penalosa commented on packages/create-cloudflare/src/helpers/command.ts:
```diff
@@ -276,3 +300,28 @@ export const listAccounts = async () => {
+ * @returns The latest compatibility date for workerd in the form "YYYY-MM-DD"
+ */
+export async function getWorkerdCompatibilityDate() {
+	return runCommand("npm info workerd dist-tags.latest", {
```
Should this run something different depending on the package manager that was used? Or is the assumption that npm will always be installed (which seems like a fair assumption)?

petebacondarwin replied:
Of course Yarn output is slightly different to npm and pnpm :sad:

petebacondarwin replied:
Added support for this and improved the date matching to accommodate yarn.
And added tests.

penalosa commented on packages/create-cloudflare/src/helpers/command.ts:
```diff
@@ -276,3 +300,28 @@ export const listAccounts = async () => {
+			const date = result.split(".")[1];
+			return `${date.slice(0, 4)}-${date.slice(4, 6)}-${date.slice(6, 8)}`;
+		},
+		fallback: () => "2023-05-18",
```
Can this fallback to today/yesterday?

petebacondarwin replied:
Not really because that might be too recent for the current version of workerd.
Alternatively we could grab the most recent workerd date...
Happy for someone to follow up with a suitable alternative to this if desired.

RamIdeas commented on packages/create-cloudflare/src/helpers/command.ts:
```diff
@@ -18,12 +18,14 @@ type Command = string | string[];
 	useSpinner?: boolean;
 	env?: NodeJS.ProcessEnv;
 	cwd?: string;
+	transform?: (output: string) => string;
+	fallback?: (error: unknown) => string;
```
I think it would be more descriptive to name these `transformOutput` and `fallbackOutput`

petebacondarwin replied:
OK

RamIdeas commented on packages/create-cloudflare/src/helpers/command.ts:
```diff
@@ -18,12 +18,14 @@ type Command = string | string[];
 type RunOptions = {
 	startText?: string;
-	doneText?: string;
+	doneText?: string | ((result: string) => string);
```
In line with my other recommendation, the param here would be better named `output`

```suggestion
	doneText?: string | ((output: string) => string);
```

It would be clearer to understand what their purpose is without having to read the source code
